### ISSUE TITLE
Empty statements not working?

### ISSUE THREAD

**Issue by burner1024** [sfall-team/sfall#349]
> empty statements in blocks are allowed: This is just a convenience to save scripters a bit of memory. Some of the macros in the fallout headers include their own semicolons while others do not. With the original compiler you had to remember which was which, and if you got it wrong the script would not compile. Now it's always safe to include your own semicolon, even if the macro already had its own.

Trying with 4.2.8.1: 
```
$ cat 1.ssl
#define my_macro display_msg("foo");

procedure start begin
  my_macro;
end
$ wine ~/bin/compile.exe -p -n -l -O2 -q 1.ssl -o 1.int
Compiling 1.ssl
[Error] <1.ssl>:5:4: expecting ';'.

*** THERE WERE ERRORS (1 of them) ***
```
Tried some earlier 4.x versions, also no go.

**Comment by NovaRain** [comment]
Removing the semicolon from one of them should work.
I tried the sslc as old as 3.3, and it still gives the same error.

**Comment by burner1024** [comment]
Sure, but the point is exactly not having to remove the semicolon.
If the readme is wrong and it never worked, then consider it a feature request.

**Comment by FakelsHub** [comment]
All macro definitions are written without `;` at the end.
Consider this a syntax rule.

**Comment by phobos2077** [comment]
I agree, this looks like a bug. Either we should add this feature or remove this message from the docs.

**Comment by NovaRain** [comment]
The first mention of this "feature" is in 2.11, but even the compile.exe from 2.11 doesn't work with the example.

**Comment by burner1024** [comment]
Is it a big problem to allow this?

**Comment by phobos2077** [comment]
> 
> 
> Is it a big problem to allow this?

I don't think so. Just a matter of tweaking the parser a bit.

**Comment by FakelsHub** [comment]
> Just a matter of tweaking the parser a bit.

Предлагаю вам этим и заняться. :)

**Comment by NovaRain** [comment]
Did some quick comparison. The major change from sslc 2.9 to 2.11 is including the mcpp preprocessor, but I failed to notice if Timeslip added extra handling for semicolon.

**Comment by FakelsHub** [comment]
Napoleon's plans are plans that are difficult and sometimes impossible to implement.
All these hacks only make it harder to understand the code, even if you rewrite everything in c++, the code will still be difficult to understand, since it is essentially divided into various small parts.

**Comment by phobos2077** [comment]
> У меня от кода того парсера голова пухнет, все очень там запутано. в общем си код там говно, и отлаживать его еще тот АД.

Я лет 5 на него не смотрел, но помню что именно парсер там супер-простой. Если нужно небольшое изменение внести - это элементарно. А вот новые фичи которые требуют много изменений - уже сложновато.

**Comment by Mingun** [comment]
А там парсер рукописный (и где он вообще)? Обычно если парсер в виде какой-то грамматики, то вносить изменения не сложно, что мелкие, что средней тяжести.

**Comment by FakelsHub** [comment]
> 
> 
> А там парсер рукописный (и где он вообще)?

[parce.c](https://github.com/FakelsHub/int-sslc/blob/master/sslc/parse.c)

**Comment by FakelsHub** [comment]
@NovaRain 
you can try replacing the second argument of the function `parseStatementInternal(p, 1);` to 0 and see how the compiler will work.
I have doubts that this didn't work well, so it was eventually disabled.

**Comment by NovaRain** [comment]
> 
> 
> @NovaRain
> you can try replacing the second argument of the function `parseStatementInternal(p, 1);` to 0 and see how the compiler will work.
> I have doubts that this didn't work well, so it was eventually disabled.

No good, for a code like this:
```pas
   for (i = 0; i < 9; i++) begin // line 447
   // do something
   end
```
The compiler will complain `[Error] <gl_0keytest.ssl>:447:17: Expected ';'` (the i in `i < 9`).

**Comment by FakelsHub** [comment]
The exception for parseWhile

**Comment by NovaRain** [comment]
> 
> 
> The exception for parseWhile

OK, I replaced the `parseStatement(p);` in `parseBlock()` and `parseIf()` but not parseWhile(), both the example script and the one using for loop pass the compilation, and the decompiled results seem normal.

**Comment by phobos2077** [comment]
Whatever you guys change, don't forget to test by re-compiling ALL original scripts (with RP scripts as well for good measure) and making sure it still works.

**Comment by FakelsHub** [comment]
I would not use the method of optional semicolon characters, and would remove this description from the documentation.
Ps. semicolon characters (can) used as helper with parsing the code.

**Comment by burner1024** [comment]
Problem is, some defines _must_ have a semicolon, while others _must not._ And as of now you have to remember which is which all the time.

**Comment by FakelsHub** [comment]
С этого момента ты должен установить нормальную IDE для разработки и не париться с запоминанием.

Как вариант сделать отдельную опцию компилятора, и тогда вся боль ляжет на того кто будет компилироваться без точки с запятой. :-)

**Comment by FakelsHub** [comment]
А по факту нужно делать совсем другую функцию,  которая бы убирала двойной сивол `;`
Но легче это сделать в скрипт редакторе чем в этом компиляторе.

**Comment by burner1024** [comment]
Какую это нормальную IDE?
Ну даже если будет подсказка, дефайны бывают в сотни строк. Это каждый раз надо мотать смотреть.
Я не вижу в чем разницу сделать опцию и сделать без опции, код ведь тот же самый?

**Comment by FakelsHub** [comment]
> Какую это нормальную IDE?
> Ну даже если будет подсказка, дефайны бывают в сотни строк. Это каждый раз надо мотать смотреть.

ну по факту никакой IDE для нашего дела не существует (есть какая-то там универсально-настраиваемая Eclipse) кроме Sfall Script Editor, который подчеркнет ошибку, покажет или перейдет к макросу.

> Я не вижу в чем разницу сделать опцию и сделать без опции, код ведь тот же самый?

с опцией это во первых страховка.
к примеру Sfall Script Editor при парсинге кода скрипта определяет конец функции/макроса по этому символу.

**Comment by burner1024** [comment]
> ну по факту никакой IDE для нашего дела не существует 

Ну вот

> к примеру Sfall Script Editor при парсинге кода скрипта определяет конец функции/макроса по этому символу.

... а значит если в макросе нет и не предолагается ";", то SSE вскрякнет. Так это в нем косяк, а не страховка.

**Comment by FakelsHub** [comment]
> Так это в нем косяк, а не страховка.

причем тут "в нем косяк", точка запятой это часть синтаксиса SSL.
_поди напиши разработчикам питона, что там их отступы к черту не нужны, тебя отправят к царю))) так и здесь._
страховка нужна от косяков компилятора, которые вы будете искать) как пользователи данной фичи

**Comment by burner1024** [comment]
Я вообще не понимаю, что ты имеешь ввиду.
С опцией так с опцией, если хоть как-то будет компилироваться пустой statement, это будет отлично.

**Comment by phobos2077** [comment]
After some investigation, we couldn't find any evidence that "empty statements" ever worked. Adding this doesn't seem to provide much benefit but may lead to some parsing issues. So I don't think we should do it.
